dicom_utils/volume.py

```diff
@@ -2,28 +2,87 @@
 # -*- coding: utf-8 -*-
 
 from abc import ABC, abstractmethod
-from typing import Sized, SupportsIndex, TypeVar
+from copy import deepcopy
+from itertools import islice
+from typing import Iterator, Optional, Sized, SupportsIndex, SupportsInt, Tuple, TypeVar, Union, overload
+
+import numpy as np
+from pydicom import Dataset
+from pydicom.encaps import encapsulate, generate_pixel_data_frame
 
 
 T = TypeVar("T", bound=SupportsIndex)
+U = TypeVar("U", bound=Dataset)
 
 
 class VolumeHandler(ABC):
     r"""Base class for classes that manipulate 3D Volumes"""
 
     @abstractmethod
+    def get_indices(self, total_frames: Optional[int]) -> Tuple[int, int, int]:
+        ...
+
+    @overload
     def __call__(self, x: T) -> T:
         ...
 
+    @overload
+    def __call__(self, x: U) -> U:
+        ...
+
+    def __call__(self, x: Union[T, U]) -> Union[T, U]:
+        if isinstance(x, Dataset):
+            return self.slice_dicom(x)
+        else:
+            return self.slice_array(x)
+
+    def slice_array(self, x: T) -> T:
+        r"""Slices an array input according to :func:`get_indices`"""
+        num_frames = len(x) if isinstance(x, Sized) else None
+        start, stop, stride = self.get_indices(num_frames)
+        result = x[start:stop:stride]  # type: ignore
```

Comment:
I think this `# type: ignore` could be removed if we leveraged a Protocol class:
```python
class SupportsGetItem(SupportsIndex, Protocol):
    def __getitem__(self: "SupportsGetItem", key: Any) -> Any:
        ...


T = TypeVar("T", bound=SupportsGetItem)
```

This does cause a type issue in `dicom.py` which could be fixed with something like:
```python
if len(dims) == 4:
    should_be_dicom = volume_handler(dcm)
    assert isinstance(should_be_dicom, Dicom)
    dcm = should_be_dicom
```